src/main/java/gyro/google/compute/FirewallResource.java

```diff
@@ -0,0 +1,546 @@
+/*
+ * Copyright 2019, Perfect Sense, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package gyro.google.compute;
+
+import com.google.api.client.googleapis.json.GoogleJsonResponseException;
+import com.google.api.services.compute.Compute;
+import com.google.api.services.compute.model.Firewall;
+import com.google.api.services.compute.model.FirewallLogConfig;
+import com.google.api.services.compute.model.Operation;
+import com.google.cloud.compute.v1.ProjectGlobalNetworkName;
+import gyro.core.GyroException;
+import gyro.core.GyroUI;
+import gyro.core.Type;
+import gyro.core.resource.Id;
+import gyro.core.resource.Output;
+import gyro.core.resource.Resource;
+import gyro.core.resource.Updatable;
+import gyro.core.scope.State;
+import gyro.core.validation.Range;
+import gyro.core.validation.Required;
+import gyro.core.validation.ValidStrings;
+import gyro.core.validation.ValidationError;
+import gyro.google.Copyable;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Creates a firewall rule.
+ *
+ * Example
+ * -------
+ *
+ * .. code-block:: gyro
+ *
+ *     google::firewall-rule firewall-rule-example
+ *         name: "firewall-rule-example"
+ *         network: $(google::network network-example-firewall-rule)
+ *         description: "firewall-rule-example-desc"
+ *         direction: "ingress"
+ *         priority: 1001
+ *
+ *         rule-type: "allow"
+ *
+ *         rule
+ *             protocol: "tcp"
+ *             ports: ["95-96", "80-85"]
+ *         end
+ *
+ *         source-tags: [
+ *             "source-tag-example"
+ *         ]
+ *     end
+ */
+@Type("firewall-rule")
+public class FirewallResource extends ComputeResource implements Copyable<Firewall> {
+    private String name;
+    private NetworkResource network;
+    private String description;
+    private String ruleType;
+    private Set<FirewallAllowDenyRule> rule;
```

Comment:
This doesn't follow the GCP data model. In their model they have an `allowed` list and `denied` list. I realize the model of those two lists is the same but we should mirror the underlying API so we're consistent.